odoo/models.py

```diff
@@ -5375,7 +5375,7 @@ def filtered_domain(self, domain):
             else:
                 (key, comparator, value) = d
                 if comparator in ('child_of', 'parent_of'):
-                    result.append(self.search([('id', 'in', self.ids), d]))
+                    result.append(self.with_context(active_test=False).search([('id', 'in', self.ids), d]))
```

Comment:
This patch is at the wrong place, as `filtered_domain` is more general than checking access rights.
Please move that patch to the caller `_filter_access_rules_python()`, which already adds `sudo()`.

Reply:
I think I left out some parts in the description of the problem. It's not really a permissions problem, but I found the problem because of them. The real problem is that the method is not consistent, the demo has been added as a test and can be replicated with odoo shell

On the other hand, I consider that adding the context in the `_filter_access_rules_python` function will be giving more load to the permissions checks, since this search occurs only when you send a tuple in the domain with the `child_of` or `parent_of` operator. 

And if the `filtered_domain` method is called from somewhere else, it should still return all records matching the specified domain, i.e. the method should not filter out what is not explicitly in the domain being sent.

Reply:
Well, the issue is that `filtered_domain()` does not discard inactive records _at the toplevel_, but it does on sublevels.

Consider a one2many field `line_ids`.  Calling `records.filtered_domain([('line_ids.amount', '>', 0)])` returns the records that have at least one _active_ line satisfying `[('amount', '>', 0)]`, unless `active_test=False` has been added in the context of `records`.  And `search()` does the same, even with hierarchy operators.  And the conversion of record rules in plain SQL does it as well (see `_filter_access_rules`.)

The design of `filtered_domain` follows the principle of being equivalent to a `search` where the "active test" is not considered at the toplevel.  I think that your proposal breaks this design principle.

To see whether things are consistent, you have to check
```python
records.filtered_domain(domain) == records.search(domain + [('id', 'in', records.ids), ('active', 'in', [True, False])])
```

Reply:
Currently is not consistent, if we have an archived partner and we apply the filtered_domain and search methods; we will see that the search will **not return this record in any case** unless, as you mention, we use the `active_test=False` context, however, if we do the same tests with the filtered_domain method, we will find that it will **return the record in all cases** except when the operator is `child_of` or `parent_of`.

If the filtered_domain follows the principle of being equivalent to search. Under this assumption, `filtered_domain` is not consistent with `search`, and we need exclude the archived records when we are filtering for all the other operators '=', '>', 'in', 'like', etc.

Note: I don't include on search ("active", "in" [True, False]) because it's not part of the "actual" domain test.
```python
partner = self.env["res.partner"].search([], limit=1)
partner.write({"active": False})
# Out: True
partner.active
# Out: False

# Inconsistency 1: filtered domain return the records
partner.filtered_domain([])
# Out: res.partner(14,)
partner.search([("id", "in", partner.ids)])
# Out: res.partner()

# Inconsistency 2: filtered domain return the records with = operator
partner.filtered_domain([("name", "=", partner.name)])
# Out: res.partner(14,)
partner.search([("id", "in", partner.ids), ("name", "=", partner.name)])
# Out: res.partner()

# Actual part of the discussion, made this 'inconsistent' with search? or made all the others 'consistent' with search?
partner.filtered_domain([("id", "child_of", partner.id)])
# Out: res.partner()
partner.search([("id", "in", partner.ids), ("id", "child_of", partner.id)])
# Out: res.partner()

# Inconsistency 3: filtered domain return the records with like operator
partner.filtered_domain([("name", "like", partner.name)])
# Out: res.partner(14,)
partner.search([("id", "in", partner.ids), ("name", "like", partner.name)])
# Out: res.partner()

# Inconsistency 4: filtered domain return the records with ilike operator
partner.filtered_domain([("name", "ilike", partner.name)])
# Out: res.partner(14,)
partner.search([("id", "in", partner.ids), ("name", "ilike", partner.name)])
# res.partner()
```

Reply:
@rco-odoo 
I have included the changes you have proposed, however, now the tests I included are failing because of the above issue, what would be the approach we should take?

Reply:
> ... what would be the approach we should take?

Change the rule, and add an explicit condition to include archived records: `('active', 'in', [True, False])`.